Feat/12/home UI

SunnnySong opened the pull request:
## 📕 Issue
Close #12 


<br/>

## 📙 작업 내역
> 구현 내용 및 작업 했던 내역
- [x] Home화면 UI
- [x] 재활용 가능한 UI Objects 생성하기
- [x] Home화면 스크롤뷰 구현하기
- [x] 구현에 편리한 extension 정의하기


<br/>

## 📘 작업 유형
- [x] 신규 기능 추가
- [ ] 버그 수정
- [ ] 리팩토링
- [ ] 문서 업데이트


<br/>

## 📸 Scree
<img src="https://github.com/seunghyunCheon/LottoDiary/assets/85678496/37353aba-ba30-4aa1-9558-c5d790568817" width="250" height="550"/>




<br/>

## 📝 PR 특이사항
> PR을 볼 때 주의깊게 봐야하거나 말하고 싶은 점
- UIScrollView를 설정했지만, View hierarchy를 띄어 보면 UIScrollView에 보라색 삼각형 경고가 띄어져 있습니다😭
    - 이 에러를 해결하고 PR을 보내기엔 너무 시간이 오래 걸릴거 같아 먼저 올려요! 혹시 해당 오류의 해결 방법을 알고 계신다면 공유 부탁드립니다,,

SunnnySong committed:
feat: HomeView 중 목표, 구매, 당첨금액 안내 StackView 구현

- UIStackView+ : addArrangedSubview를 할때 여러 subviews를 한번에 추가할 수 있는 함수 구현
- MoneyInformationStackView : 목표, 구매, 당첨금액을 보여주는 가로 StackView 구현

SunnnySong committed:
feat: 기본적인 HomeScene UI 구현

SunnnySong committed:
Merge branch 'develop' into feat/12/Home-UI

SunnnySong committed:
chore: HomeView 의 background 색상 변경

SunnnySong committed:
feat: ScrollView 추가

SunnnySong committed:
feat: MoneyInformationView 관련 UI 수정

SunnnySong committed:
refactor: View로 분리했던 UIObjects -> UIViewController로 이동

SunnnySong committed:
Merge branch 'develop' into feat/12/Home-UI

SunnnySong committed:
feat: home viewController 가로 양 옆으로 여백 추가

SunnnySong committed:
feat: attributed를 사용한 Text 관련 속성 설정

SunnnySong committed:
chore: AppCoordinator 시작 flow 변경

SunnnySong committed:
chore: 필요없는 파일 삭제

SunnnySong committed:
chore: enum 범위 기존 코드대로 변경

SunnnySong committed:
chore: enum 범위 기존 코드대로 변경

SunnnySong committed:
Merge remote-tracking branch 'origin/feat/12/Home-UI' into feat/12/Home-UI

seunghyunCheon commented on LottoDairy/Common/GmarketSansLabel.swift:
```diff
@@ -0,0 +1,28 @@
+         size: UIFont.Size,
+         weight: UIFont.CustomWeight) {
+        super.init(frame: .zero)
+
```
추후 다이나믹 타입 대응을 위해 `adjustsFontForContentSizeCategory` 속성을 추가하는 건 어떨까요?

seunghyunCheon commented on LottoDairy/HomeScene/HomeViewController.swift:
```diff
@@ -8,12 +8,228 @@
+        let height = view.frame.height * 0.35
+        let nickNameViewHeight: CGFloat = height * 0.127
+        let explanationLabelGap: CGFloat = height * 0.07
+        let explanationLabelHeight: CGFloat = height * 0.27
```
숫자 수식도 나중에 알아보기 쉽게 상수로 빼도 좋을 것 같아요!

seunghyunCheon commented on LottoDairy/HomeScene/HomeViewController.swift:
```diff
@@ -8,12 +8,228 @@
+    }
+
+    private func setupMoneyInformationStackView() -> UIStackView {
+        let moneyInformationStackView: UIStackView = {
+            let stackView = UIStackView()
+            stackView.axis = .vertical
+            stackView.distribution = .fillEqually
+            stackView.translatesAutoresizingMaskIntoConstraints = false
+            stackView.backgroundColor = .designSystem(.gray2D2B35)
+            stackView.clipsToBounds = true
+            stackView.layer.cornerRadius = Constant.cornerRadius
+            return stackView
+        }()
```
저는 이런 스택뷰를 만들어야한다면 UI쪽에 올려서 작성할 것 같아요. UI와 동작하는 부분을 최대한 분리시키고 싶기 때문인데 이것저것 찾아보니까 써니처럼 많이 작성하는 것 같더라구요! 취향차이인 것 같아요. 이것도 깔끔해서 좋습니다 👍

seunghyunCheon commented on LottoDairy/HomeScene/HomeViewController.swift:
```diff
@@ -8,12 +8,228 @@
+
+    // MARK: Lifecycle
     override func viewDidLoad() {
         super.viewDidLoad()
 
-        view.backgroundColor = .green
+        configureView()
+        setupScrollView()
+        setupContentView()
+        configureContentView()
+    }
```
viewDidLoad 추상화레벨이 맞아서 좋아요 👍
다만 configureView가 맨처음에 호출되지만 맨 마지막에 정의되어있어서 찾기가 힘들었습니다. 저는 순서대로 쓰는데 의도하신 것이 있을까요?

seunghyunCheon commented on LottoDairy/HomeScene/HomeViewController.swift:
```diff
@@ -8,12 +8,228 @@
+        imageView.clipsToBounds = true
+        imageView.layer.cornerRadius = Constant.cornerRadius
+        imageView.translatesAutoresizingMaskIntoConstraints = false
+        return imageView
```
`return`개행에 대한 부분은 금요일 회의에 컨벤션 정리하면서 정하면 좋을 것 같아요!

seunghyunCheon commented on LottoDairy/HomeScene/HomeViewController.swift:
```diff
@@ -8,12 +8,228 @@
+}
+
+// MARK: Extension
+extension HomeViewController {
+
+    private enum SystemName: String {
+        case setting = "gearshape"
+        case photo = "photo"
+
+        var image: UIImage? {
+            return UIImage(systemName: self.rawValue)
+        }
+    }
+
+    private enum Constant {
+        static let cornerRadius: CGFloat = 15
+    }
+
+    private enum StringLiteral {
+        static let goalTitle = "목표"
+        static let buyTitle = "구매 금액"
+        static let winTitle = "당첨 금액"
+        static let imageTitle = "이 돈이면"
     }
 }
```
깔끔하네요! 저도 참고할게요! 👍

seunghyunCheon commented on LottoDairy/HomeScene/Views/DoubleLabelView.swift:
```diff
@@ -0,0 +1,141 @@
+
+import UIKit
+
+class DoubleLabelView: UIView {
```
`final`이 빠져있습니다.

seunghyunCheon commented on LottoDairy/HomeScene/Views/DoubleLabelView.swift:
```diff
@@ -0,0 +1,141 @@
+    convenience init(month: String, percent: String) {
+        self.init(frame: .zero)
+
+        firstLabel.attributedText(first: "\(month)월", second: "동안", secondFontSize: .callout)
+        secondLabel.attributedText(first: "목표치의 \(percent)%", second: "를 사용하셨습니다.", secondFontSize: .callout)
+
```
저는 `attributedString`을 이렇게 작성했었는데 사람마다 구현 방법이 다른 것이 신기하네요 ㅎㅎ
```swift
// NSAttributedString 사용예시
class ViewController: UIViewController {

    @IBOutlet weak var attributedStringLabel: UILabel!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        attributedStringLabel.attributedText = NSMutableAttributedString()
            .redHighLight("빨간색강조")
            .bold(string: "굵은 글씨", fontSize: 14)
    }
}

extension NSMutableAttributedString {
    var fontSize: CGFloat {
        return 14
    }
    
    var boldFont: UIFont {
        return UIFont.boldSystemFont(ofSize: fontSize)
    }
    
    var normalFont: UIFont {
        return UIFont.boldSystemFont(ofSize: fontSize)
    }
    
    func bold(string: String, fontSize: CGFloat) -> NSMutableAttributedString {
        let attributes: [NSAttributedString.Key: Any] = [.font: boldFont]
        self.append(NSAttributedString(string: string, attributes: attributes))
        return self
    }
    
    func redHighLight(_ value: String) -> NSMutableAttributedString {
        let attributes: [NSAttributedString.Key: Any] = [
            .font: normalFont,
            .foregroundColor: UIColor.white, // 글자 색
            .backgroundColor: UIColor.red // 배경 색
        ]
        self.append(NSAttributedString(string: value, attributes: attributes))
        return self
    }
}
```
코드이해가 잘 되었습니다.

seunghyunCheon commented on LottoDairy/HomeScene/Views/DoubleLabelView.swift:
```diff
@@ -0,0 +1,141 @@
+    }
+
+    private func setupVerticalView() {
+        let gab: CGFloat = 10
```
오타가 난 것 같아요.

seunghyunCheon commented on LottoDairy/HomeScene/HomeViewController.swift:
```diff
@@ -8,12 +8,228 @@
+            scrollView.leadingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.leadingAnchor),
+            scrollView.trailingAnchor.constraint(equalTo: view.safeAreaLayoutGuide.trailingAnchor),
+
+            contentView.topAnchor.constraint(equalTo: scrollView.contentLayoutGuide.topAnchor),
+            contentView.bottomAnchor.constraint(equalTo: scrollView.contentLayoutGuide.bottomAnchor),
+            contentView.widthAnchor.constraint(equalTo: scrollView.widthAnchor)
+        ])
```
이 곳에서 contentView의 leading과 trailing 제약을 contentLayoutGuide에 맞춰주니까 스크롤뷰 이슈가 해결되었습니다.

seunghyunCheon reviewed (commented):
코드 잘 봤습니다! 😄 `DoubleLabelView`를 통해 뷰컨을 가볍게 만들어서 좋다고 느꼈습니다! 저도 거의 모든 것을 커스텀으로 빼봐야겠네요 ㅎㅎ

스크롤 뷰 이슈사항은 코멘트에 남겨두었습니다. 다만 DoubleLabelView에 관한 이슈가 존재해서 이 부분 해결하면 좋을 것 같아요!

그리고 `moneyInforamtionStackView`의 디자인은 아래 사진처럼 하는 것은 어떨까요?

<img src="https://github.com/seunghyunCheon/LottoDiary/assets/70146658/ae3466b7-16b6-48d2-9cf9-95c13bfa2e40" width="200"/><br/>
왼쪽 정렬로 맞추게 되면 금액이나 subLabel의 수가 많아지더라도 유연하게 될 것 같다고 생각합니다.

마지막으로 iphone SE기종으로 시뮬레이터를 돌려봤을 때 autolayout이 어색한 부분이 있어요. 이 부분 확인해주시면 좋을 것 같아요.
<img width="373" alt="image" src="https://github.com/seunghyunCheon/LottoDiary/assets/70146658/b9edb04f-2dab-4d81-bc2e-7babfb113737">

SunnnySong committed:
review: 다이나믹 타입 적용을 위한 속성 설정

- 다이나믹 타입 적용을 위해 adjustsFontForContentSizeCategory 속성 true 설정

SunnnySong committed:
review: return 컨벤션 수정

SunnnySong committed:
review: 함수 호출과 함수 정의 간 순서 맞추기

SunnnySong committed:
review: final 키워드 추가

SunnnySong committed:
review: 오타 수정

SunnnySong committed:
review: ScrollView 관련 이슈 수정

SunnnySong committed:
refactor: 어색한 DoubleLabelView layout 수정하기

- UILabel은 왠만하면 height 고정하지 않기
- 내장되어있는 height를 사용해야 글씨 크기가 달라질 경우(다이나믹과 같은) label의 글씨가 잘리지 않고 보여질 수 있다.

SunnnySong committed:
refactor: 어색한 DoubleLabelView layout 수정하기

- UILabel은 왠만하면 height 고정하지 않기
- 내장되어있는 height를 사용해야 글씨 크기가 달라질 경우(다이나믹과 같은) label의 글씨가 잘리지 않고 보여질 수 있다.

SunnnySong committed:
refactor: MoneyInformationStackView 요소 위치 변경

SunnnySong committed:
Merge remote-tracking branch 'origin/feat/12/Home-UI' into feat/12/Home-UI

SunnnySong committed:
feat: moneyHorizontalStackView의 alignment 변경

- label의 폰트 사이즈에 따라 라벨 크기가 유동적으로 변경될 수 있도록 UIStackView로 설정했는데 라벨 크기가 고정되는 문제 발생
- moneyHorizontalStackView의 alignment 를 설정하니 문제 해결!
